sklearn/neighbors/base.py

```diff
@@ -391,20 +400,45 @@ def kneighbors_graph(self, X, n_neighbors=None,
 
         # construct CSR matrix representation of the k-NN graph
         if mode == 'connectivity':
-            A_data = np.ones((n_samples1, n_neighbors))
-            A_ind = self.kneighbors(X, n_neighbors, return_distance=False)
+            A_data = np.ones(n_samples1 * n_neighbors)
+
+            # If the query data in kneighbors_graph is the same as the indexed
+            # data, then we need one extra neighbor, since the first neigbor
+            # is the sample itself. The same argument holds true for mode=distance
+            if not new_graph:
+                A_ind = self.kneighbors(X, n_neighbors + 1, return_distance=False)
+                # First column coressponds to the diagonal
```

Comment:
But what is wrong with the current behavior, it treats the duplicates as new samples.

    In [14]: from sklearn.neighbors import NearestNeighbors
    In [15]: X = np.random.rand(5, 5)
    In [16]: X[1], X[2] = X[0], X[0]
    In [17]: nn = NearestNeighbors(n_neighbors=2).fit(X).kneighbors_graph().A
    In [18]: nn
    Out[18]: 
    array([[ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  1.,  1.,  0.,  0.],
           [ 0.,  0.,  1.,  1.,  0.]])